### PR TITLE
Metal: MTLComputeCommandEncoder Dependency Structs

### PR DESCRIPTION
**Pull request by IsaacMarovitz**
## Description

Adds serialisation code for all structs and enums that `MTLComputeCommandEncoder` and its dependencies rely on.

### CODE REVIEW

File: renderdoc/driver/metal/metal_types.h
Comment (Zorro666) [inline renderdoc/driver/metal/metal_types.h]: I am interested in your system setup (I have only ever been able to compare/use Metal development on my own one system. For me the Metal files are in MacOSX14.2.sdk i.e.

Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/System/Library/Frameworks/Metal.framework/Headers/MTLStageInputOutputDescriptor.h

but for you you have Metal files in MacOSX12.1.sdk and MacOSX14.2.sdk

I suspect the files are quite similar in both places. The reason I mention it I found it a bit tricky to keep track of any MaxOS changes to Metal header files. I now keep a local copy of the Metal header files at a certain OS version i.e. currently the project is on MacOSX 14.0 then when a new Xcode or MacOS is released with new Metal version. I copy the headers over into a new folder and manually compare the Metal headers to find any changes and hand copy the changes over into the wrapped classes.

I checked locally and for MTLStageInputOutputDescriptor.h MacOSX14.0 and MacOSX12.1 versions are basically identical.

File: renderdoc/driver/metal/metal_types.h
Comment (Zorro666) [inline renderdoc/driver/metal/metal_types.h:494]: This struct does not look to be complete - for me it is something like this in the ObjC
```
@interface MTLComputePipelineDescriptor : NSObject <NSCopying>

@property (nullable, copy, nonatomic) NSString *label;
@property (nullable, readwrite, nonatomic, strong) id <MTLFunction> computeFunction;
@property (readwrite, nonatomic) BOOL threadGroupSizeIsMultipleOfThreadExecutionWidth;
@property (readwrite, nonatomic) NSUInteger maxTotalThreadsPerThreadgroup;
@property (nullable, copy, nonatomic) MTLStageInputOutputDescriptor *stageInputDescriptor;
@property (readonly) MTLPipelineBufferDescriptorArray *buffers;
@property (readwrite, nonatomic) BOOL supportIndirectCommandBuffers;
@property (readwrite, nullable, nonatomic, copy) NSArray<id<MTLDynamicLibrary>>* insertLibraries;
@property (readwrite, nonnull, nonatomic, copy) NSArray<id<MTLDynamicLibrary>>* preloadedLibraries;
@property (readwrite, nullable, nonatomic, copy) NSArray<id<MTLBinaryArchive>> *binaryArchives;
@property (nullable, copy, nonatomic) MTLLinkedFunctions *linkedFunctions;
@property (readwrite, nonatomic) BOOL supportAddingBinaryFunctions;
@property (readwrite, nonatomic) NSUInteger maxCallStackDepth;

@end
```

**Review by Zorro666 — CHANGES_REQUESTED** [review]
Looks like there is some missing serialisation : having a working capture and replay prototype would be a great way to check the serialisation is working

File: renderdoc/driver/metal/metal_types.h
Comment (IsaacMarovitz) [reply renderdoc/driver/metal/metal_types.h]: Oh this was probably a typo, as I only have the 14.2 SDKs locally, I'll correct it.

File: renderdoc/driver/metal/metal_types.h
Comment (IsaacMarovitz) [reply renderdoc/driver/metal/metal_types.h:494]: Added missing fields

File: renderdoc/driver/metal/metal_serialise.cpp
Comment (Zorro666) [inline renderdoc/driver/metal/metal_serialise.cpp:403]: No need to change anything but a comment for the future : don't need the deprecated here as this method is about serialising the structure and not about what the data means. Also it might be misunderstood if the field is deprecated in the RDMTL structure as opposed to deprecated in the ObjC structure.

File: renderdoc/driver/metal/metal_types.h
Comment (Zorro666) [inline renderdoc/driver/metal/metal_types.h:525]: No need to change just explaining a development principle I have been using. 

The documentation isn't fully clear and I don't trust documentation to match code behaviour as a rule.
In the past I have called the ObjC create methods i.e. `computePassDescriptor` to see what defaults it makes.
The documentation does imply `MTL::DispatchTypeSerial` is the default

File: renderdoc/driver/metal/metal_types.h
Comment (Zorro666) [inline renderdoc/driver/metal/metal_types.h]: Please create a default `ComputePIpelineDescriptor` and look at the default values to check them  : I think the default for this property might be zero from reading the documentation.

 @abstract Optional property. Set the maxTotalThreadsPerThreadgroup. If it is not set, returns zero

File: renderdoc/driver/metal/metal_stringise.cpp
Comment (Zorro666) [inline renderdoc/driver/metal/metal_stringise.cpp:1195]: A note in the 14.0 Metal ObjC header file there are these enum values
```
MTLAttributeFormatFloatRG11B10,
MTLAttributeFormatFloatRGB9E5,
```
They aren't in the Metal CPP file because I haven't updated Metal CPP to SDK 14.x and update the serialisation.

Please could you add these two new enum values here, commented out and with a TODO saying when Metal CPP is updated to 14.x

File: renderdoc/driver/metal/metal_types.cpp
Comment (Zorro666) [inline renderdoc/driver/metal/metal_types.cpp]: This looks like a typo and should be 

    : startOfEncoderSampleIndex(objc-> startOfEncoderSampleIndex()),

**Review by Zorro666 — CHANGES_REQUESTED** [review]
Thank you for the updates. It is looking very good.
There are a couple of minor things that would be good to change before the PR is merged.

Looking forward to seeing the replay work for compute (would need to be done in the WIP branch with replay support).

I will be continuing to upstream the Metal capture and replay work which will make things simpler for development in the future

File: renderdoc/driver/metal/metal_types.h
Comment (IsaacMarovitz) [reply renderdoc/driver/metal/metal_types.h]: You're right, it defaults to 0

File: renderdoc/driver/metal/metal_types.h
Comment (IsaacMarovitz) [reply renderdoc/driver/metal/metal_types.h:525]: I checked and `DispatchTypeSerial` is correct.

**Review by Zorro666 — APPROVED** [review]
Thank you for the iterations and for the Metal PR.
I will merge the PR after the next RenderDoc release is completed which should be next week.
I will also upstream an updated metal-cpp and continue to upstream the current metal capture and replay work

**Pull request merged**
